### PR TITLE
[Plugin|plugins] Consolidate command collection API methods

### PR DESCRIPTION
**Pull request by TurboTurtle**
Removed `call_ext_prog()`, `check_ext_prog()`, and
`get_command_output()` in favor of consolidating Plugin command
collection into two public methods: `add_cmd_output()` and
`get_cmd_output_now()`.

- `add_cmd_output()` remains unchanged and is the interface to use when
plugins should collect command output during their collect() phase.

- `get_cmd_output_now()` will now return the content of a command's
execution as well as the filename within the sos archive, instead of
just returning the filename. This should help streamline plugin code
that runs a command and then iterates over that output.

The `get_cmd_output_now()` has a `save` keyword arg (default True), that
if set to False will skip writing the command output to the archive.

Also updates the plugins to no longer use `get_command_output()`,
`call_ext_prog()`, or `check_ext_prog()` - instead only using
`get_cmd_output_now()`.

For the vast majority of plugins, these calls were able to be changed
in-place with little to no further alteration of plugin code outside of
no longer making indenpendent open() calls to files created, as the
content of those files is now directly available. Exceptions are noted
below.

For the dlm, gfs2, and kvm plugins the only use of get_command_output()
was to mount /sys/kernel/debugfs which in modern distributions is no
longer needed, so those plugins have had that functionality entirely
removed instead of being ported to the new `get_cmd_output_now()`.

For the networking plugin, many of the checks are handled by
SoSPredicates, and so those command collections now use Predicates
rather than being gated by individual `get_cmd_output_now()` calls.

---
Please place an 'X' inside each '[]' to confirm you adhere to our [Contributor Guidelines](https://github.com/sosreport/sos/wiki/Contribution-Guidelines)

- [x] Is the commit message split over multiple lines and hard-wrapped at 72 characters?
- [x] Is the subject and message clear and concise?
- [x] Does the subject start with **[plugin_name]** if submitting a plugin patch or a **[section_name]** if part of the core sosreport code?
- [x] Does the commit contain a **Signed-off-by: First Lastname <email@example.com>**?

### CODE REVIEW

**Comment by TurboTurtle** [comment]
This is a continuation from the conversation in #1707.

Also, it looks like this should address #1754 as a bonus, as `add_cmd_output()` eventually calls `_get_cmd_output_now()` which is where the docstring has been expanded.

File: sos/plugins/abrt.py [hunk 1]
```diff
@@ -24,16 +24,15 @@ class Abrt(Plugin, RedHatPlugin):
...
-            if line.startswith("Directory:"):
-                self.add_cmd_output("abrt-cli info -d '%s'" % line.split()[1])
-
     def setup(self):
         self.add_cmd_output("abrt-cli status")
-        list_file = self.get_cmd_output_now("abrt-cli list")
-        if self.get_option('detailed') and list_file:
-            self.info_detailed(list_file)
+        abrt_list = self.get_cmd_output_now("abrt-cli list")
+        if self.get_option("detailed") and abrt_list:
```
Comment (pmoravec) [inline sos/plugins/abrt.py]: Technically, abrt_list (as a return value of `get_cmd_output_now`) will always be a dict, so `and abrt_list` will always pass. We have few options here:

- test for `abrt_list["status"] == 0`
- test for `abrt_list["output"]` - that is bit redundant due to the further iteration over output splitlines
- skip this test

I am inclined to skip the test.

File: sos/plugins/block.py [hunk 2]
```diff
@@ -69,10 +53,12 @@ def setup(self):
                     "fdisk -l %s" % disk_path
                 ])
 
-        lsblk_file = self.get_cmd_output_now("lsblk -f -a -l")
+        lsblk = self.get_cmd_output_now("lsblk -f -a -l")
         # for LUKS devices, collect cryptsetup luksDump
-        if lsblk_file:
-            for dev in self.get_luks_devices(lsblk_file):
-                self.add_cmd_output('cryptsetup luksDump /dev/%s' % dev)
+        if lsblk['status'] == 0:
```
Comment (pmoravec) [inline sos/plugins/block.py]: See my comment in `abrt` - ideally, we shall be coherent here (no strong prferences, either testing for zero status or skipping such a test is fine to me).

File: sos/plugins/crio.py [hunk 3]
```diff
@@ -84,7 +84,7 @@ def setup(self):
 
     def _get_crio_list(self, cmd):
         ret = []
-        result = self.get_command_output(cmd)
+        result = self.get_cmd_output_now(cmd, save=False)
         if result['status'] == 0:
```
Comment (pmoravec) [inline sos/plugins/crio.py]: ditto here (test status or skip the test)

File: sos/plugins/dlm.py [hunk 4]
```diff
@@ -37,36 +33,17 @@ def setup(self):
...
-    def postproc(self):
-        if self._debugfs_cleanup and os.path.ismount(self.debugfs_path):
-            r = self.call_ext_prog("umount %s" % self.debugfs_path)
-            if r["status"] != 0:
-                self._log_error("could not unmount %s" % self.debugfs_path)
-
-        return
+        dlm_tool = "dlm_tool ls"
+        result = self.get_cmd_output_now(dlm_tool)
+        if result["status"] != 0:
```
Comment (pmoravec) [inline sos/plugins/dlm.py]: ditto here (test status or skip the test)

File: sos/plugins/dnf.py [hunk 5]
```diff
@@ -57,11 +57,11 @@ def setup(self):
...
 
-        if self.get_option("history"):
+        if self.get_option("history") and not self.get_option("history-info"):
             self.add_cmd_output("dnf history")
 
         if self.get_option("history-info"):
-            history = self.call_ext_prog("dnf history")
+            history = self.get_cmd_output_now("dnf history")
             transactions = -1
             if history['output']:
```
Comment (pmoravec) [inline sos/plugins/dnf.py]: Although it is not directly relevant to the change in API, could we drop this test? IMHO iterating over (potentially empty) splitlines is sufficient..

File: sos/plugins/docker.py [hunk 6]
```diff
@@ -104,7 +103,7 @@ def setup(self):
 
     def _get_docker_list(self, cmd):
         ret = []
-        result = self.get_command_output(cmd)
+        result = self.get_cmd_output_now(cmd, save=False)
         if result['status'] == 0:
```
Comment (pmoravec) [inline sos/plugins/docker.py]: ditto here (test for status or no test)

File: sos/plugins/gfs2.py [hunk 7]
```diff
@@ -36,24 +32,6 @@ def setup(self):
         ])
 
         if self.get_option("gfs2lockdump"):
-            if self._mount_debug():
```
Comment (pmoravec) [inline sos/plugins/gfs2.py]: Wasn't there some rationale to collect `/sys/kernel/debug/gfs2/*` _only_ under some further condition (i.e. `_mount_debug`)?

File: sos/plugins/gfs2.py [hunk 8]
```diff
@@ -36,24 +32,6 @@ def setup(self):
...
-            r = self.call_ext_prog("mount -t debugfs debugfs %s"
-                                   % self.debugfs_path)
-            if r["status"] != 0:
-                self._log_error("debugfs not mounted and mount attempt failed")
-                self._debugfs_cleanup = False
-        return os.path.ismount(self.debugfs_path)
-
-    def postproc(self):
-        if self._debugfs_cleanup and os.path.ismount(self.debugfs_path):
-            r = self.call_ext_prog("umount %s" % self.debugfs_path)
```
Comment (pmoravec) [inline sos/plugins/gfs2.py]: Is it really ok not to umount (something we probably mounted)?

File: sos/plugins/gluster.py [hunk 9]
```diff
@@ -108,14 +94,17 @@ def setup(self):
             else:
                 self.soslog.warn("Unable to generate statedumps, no such "
                                  "directory: %s" % self.statedump_dir)
-            state = self.get_command_output("gluster get-state")
+            state = self.get_cmd_output_now("gluster get-state")
             if state['status'] == 0:
```
Comment (pmoravec) [inline sos/plugins/gluster.py]: ditto here

File: sos/plugins/gluster.py [hunk 10]
```diff
@@ -108,14 +94,17 @@ def setup(self):
...
                 state_file = state['output'].split()[-1]
                 self.add_copy_spec(state_file)
 
-        volume_file = self.get_cmd_output_now("gluster volume info")
-        if volume_file:
-            for volname in self.get_volume_names(volume_file):
+        volume_cmd = self.get_cmd_output_now("gluster volume info")
+        if volume_cmd['status'] == 0:
```
Comment (pmoravec) [inline sos/plugins/gluster.py]: ditto here

File: sos/plugins/kata_containers.py [hunk 11]
```diff
@@ -37,7 +37,7 @@ def setup(self):
 
         # obtain a list of config files by asking the runtime
         cmd = 'kata-runtime --kata-show-default-config-paths'
-        configs = self.get_command_output(cmd)
+        configs = self.get_cmd_output_now(cmd)
         if configs and configs['status']:
```
Comment (pmoravec) [inline sos/plugins/kata_containers.py]: "`configs`" is never evaluated to `False`.

File: sos/plugins/kvm.py [hunk 12]
```diff
@@ -33,19 +30,7 @@ def setup(self):
             "/sys/module/kvm_amd/srcversion",
             "/sys/module/ksm/srcversion"
         ])
-        if not os.path.ismount(self.debugfs_path):
```
Comment (pmoravec) [inline sos/plugins/kvm.py]: Same comments like to `gfs2` plugin - is it OK in skipping this test?

File: sos/plugins/openstack_glance.py [hunk 13]
```diff
@@ -76,7 +72,10 @@ def setup(self):
 
     def running_in_container(self):
         for runtime in ["docker", "podman"]:
-            container_status = self.get_command_output(runtime + " ps")
+            container_status = self.get_cmd_output(
```
Comment (pmoravec) [inline sos/plugins/openstack_glance.py]: to call `self.get_cmd_output_now` instead? :)

File: sos/plugins/pcp.py [hunk 14]
```diff
@@ -148,10 +148,8 @@ def setup(self):
             os.path.join(self.pcp_log_dir, '*/*/config*')
         ])
 
-        # Need to get the current status of the PCP infrastructure
-        self.add_cmd_output("pcp")
         # Collect a summary for the current day
-        res = self.get_command_output('pcp')
+        res = self.get_cmd_output_now('pcp')
```
Comment (pmoravec) [inline sos/plugins/pcp.py]: with `save=False`?

File: sos/plugins/sas3ircu.py [hunk 15]
```diff
@@ -23,11 +23,9 @@ class SAS3ircu(Plugin, RedHatPlugin, DebianPlugin, UbuntuPlugin):
     def setup(self):
 
         # get list of adapters
-        result = self.call_ext_prog("sas3ircu list", timeout=5)
+        result = self.get_cmd_output_now("sas3ircu list", timeout=5)
```
Comment (pmoravec) [inline sos/plugins/sas3ircu.py]: with `save=False`?

File: sos/plugins/selinux.py [hunk 16]
```diff
@@ -27,7 +27,7 @@ def setup(self):
         ])
         self.add_cmd_output('sestatus')
 
-        state = self.get_command_output('getenforce')['output']
+        state = self.get_cmd_output_now('getenforce')['output']
```
Comment (pmoravec) [inline sos/plugins/selinux.py]: We will newly collect/store output of this command - a duplicate info resent in other cmd output(s), but it simplifies to find the info. Not sure if worth having `save=False` or not.

File: sos/plugins/teamd.py [hunk 17]
```diff
@@ -21,7 +21,7 @@ class Teamd(Plugin, RedHatPlugin, DebianPlugin, UbuntuPlugin):
 
     def _get_team_interfaces(self):
         teams = []
-        ip_result = self.get_command_output("ip -o link")
+        ip_result = self.get_cmd_output_now("ip -o link")
```
Comment (pmoravec) [inline sos/plugins/teamd.py]: with `save=False`?

File: sos/plugins/vdsm.py [hunk 18]
```diff
@@ -114,7 +114,7 @@ def setup(self):
         ])
 
         try:
-            res = self.call_ext_prog(
+            res = self.get_cmd_output_now(
```
Comment (pmoravec) [inline sos/plugins/vdsm.py]: with `save=False`?

File: sos/plugins/vdsm.py [hunk 19]
```diff
@@ -130,7 +130,7 @@ def setup(self):
             )
 
         try:
-            res = self.call_ext_prog('vdsm-client Host getStorageDomains')
+            res = self.get_cmd_output_now('vdsm-client Host getStorageDomains')
```
Comment (pmoravec) [inline sos/plugins/vdsm.py]: with `save=False`?

File: sos/plugins/vhostmd.py [hunk 20]
```diff
@@ -20,7 +20,7 @@ class vhostmd(Plugin, RedHatPlugin):
     packages = ['virt-what']
 
     def setup(self):
-        vw = self.get_command_output("virt-what")['output'].splitlines()
+        vw = self.get_cmd_output_now("virt-what")['output'].splitlines()
```
Comment (pmoravec) [inline sos/plugins/vhostmd.py]: with `save=False`?

File: sos/plugins/virsh.py [hunk 21]
```diff
@@ -43,8 +43,7 @@ def setup(self):
 
         # get network, pool and nwfilter elements
         for k in ['net', 'nwfilter', 'pool']:
-            self.add_cmd_output('%s %s-list' % (cmd, k))
-            k_list = self.get_command_output('%s %s-list' % (cmd, k))
+            k_list = self.get_cmd_output_now('%s %s-list' % (cmd, k))
```
Comment (pmoravec) [inline sos/plugins/virsh.py]: with `save=False`?

File: sos/plugins/virsh.py [hunk 22]
```diff
@@ -55,7 +54,7 @@ def setup(self):
 
         # cycle through the VMs/domains list, ignore 2 header lines and latest
         # empty line, and dumpxml domain name in 2nd column
-        domains_output = self.get_command_output('%s list --all' % cmd)
+        domains_output = self.get_cmd_output_now('%s list --all' % cmd)
```
Comment (pmoravec) [inline sos/plugins/virsh.py]: with `save=False`?

File: sos/plugins/xen.py [hunk 23]
```diff
@@ -37,7 +42,7 @@ def check_enabled(self):
         return (self.determine_xen_host() == "baremetal")
 
     def is_running_xenstored(self):
-        xs_pid = self.call_ext_prog("pidof xenstored")['output']
+        xs_pid = self.get_cmd_output_now("pidof xenstored")['output']
```
Comment (pmoravec) [inline sos/plugins/xen.py]: with `save=False`?

File: sos/plugins/yum.py [hunk 24]
```diff
@@ -99,7 +99,7 @@ def setup(self):
             # RHEL6+ alternative for this whole function:
             # self.add_cmd_output("yum-debug-dump '%s'"
             # % os.path.join(self.commons['dstroot'],"yum-debug-dump"))
-            r = self.call_ext_prog("yum-debug-dump")
+            r = self.get_cmd_output_now("yum-debug-dump")
```
Comment (pmoravec) [inline sos/plugins/yum.py]: with `save=False`?

File: sos/plugins/zfs.py [hunk 25]
```diff
@@ -27,7 +27,7 @@ def setup(self):
             "zpool status -vx"
         ])
 
-        zpools = self.call_ext_prog("zpool list -H -o name")
+        zpools = self.get_cmd_output_now("zpool list -H -o name")
```
Comment (pmoravec) [inline sos/plugins/zfs.py]: with `save=False`? (though this might be interesting to have..? (dont know zfs..))

File: sos/plugins/__init__.py [hunk 26]
```diff
@@ -1112,59 +1061,123 @@ def _get_cmd_output_now(self, cmd, suggest_filename=None,
...
-                                         sizelimit=sizelimit)
+        result = sos_get_command_output(
+            cmd, timeout=timeout, stderr=stderr, chroot=root,
+            chdir=runat, env=env, binary=binary, sizelimit=sizelimit,
+            poller=self.check_timeout
+        )
+
+        if result['status'] == 124:
+            self._log_warn(
+                "command '%s' timed out after %ds" % (prog, timeout)
```
Comment (pmoravec) [inline sos/plugins/__init__.py]: copy&paste typo:  s/prog/cmd

**Comment by pmoravec** [comment]
In general, I like the consolidation. Good work.

We are not coherent in calling `if something["status"] == 0:` - usually such a test is redundant (but dont harm), but in few cases it is afaik really needed. Not sure if it makes sense to unify the approach here.

Several other plugins needs to be updated as well - see their backtraces (usually due to different return value of `add_cmd_output` (not file name but dict)):

    composer-plugin-errors.txt
    Traceback (most recent call last):
      File "/root/sos-master-1807/sos/sosreport.py", line 964, in setup
        plug.setup()
      File "/root/sos-master-1807/sos/plugins/composer.py", line 30, in setup
        blueprints = self._get_entries("composer-cli blueprints list")
      File "/root/sos-master-1807/sos/plugins/composer.py", line 17, in _get_entries
        with open(ent_file, "r") as ents:
    TypeError: coercing to Unicode: need string or buffer, dict found
    
    kpatch-plugin-errors.txt
    Traceback (most recent call last):
      File "/root/sos-master-1807/sos/sosreport.py", line 964, in setup
        plug.setup()
      File "/root/sos-master-1807/sos/plugins/kpatch.py", line 27, in setup
        kpatches = open(kpatch_list, "r").read().splitlines()
    TypeError: coercing to Unicode: need string or buffer, dict found
    
    openstack_keystone-plugin-errors.txt
    Traceback (most recent call last):
      File "/root/sos-master-1807/sos/sosreport.py", line 964, in setup
        plug.setup()
      File "/root/sos-master-1807/sos/plugins/openstack_keystone.py", line 117, in setup
        super(RedHatKeystone, self).setup()
      File "/root/sos-master-1807/sos/plugins/openstack_keystone.py", line 55, in setup
        not(os.path.isdir(self.domain_config_dir)):
      File "/usr/lib64/python2.7/genericpath.py", line 41, in isdir
        st = os.stat(s)
    TypeError: coercing to Unicode: need string or buffer, dict found
    
    Traceback (most recent call last):
      File "/root/sos-master-1807/sos/sosreport.py", line 1167, in postproc
        plug.postproc()
      File "/root/sos-master-1807/sos/plugins/openstack_keystone.py", line 100, in postproc
        r"((?m)^\s*(%s)\s*=\s*)(.*)", r"\1********")
      File "/root/sos-master-1807/sos/plugins/__init__.py", line 595, in do_path_regex_sub
        pathexp = re.compile(pathexp)
      File "/usr/lib64/python2.7/re.py", line 190, in compile
        return _compile(pattern, flags)
      File "/usr/lib64/python2.7/re.py", line 229, in _compile
        p = _cache.get(cachekey)
    TypeError: unhashable type: 'dict'
    
    pcp-plugin-errors.txt
    Traceback (most recent call last):
      File "/root/sos-master-1807/sos/sosreport.py", line 964, in setup
        plug.setup()
      File "/root/sos-master-1807/sos/plugins/pcp.py", line 132, in setup
        for line in open(pmlogger_ls).read().splitlines():
    TypeError: coercing to Unicode: need string or buffer, dict found
    
    sapnw-plugin-errors.txt
    Traceback (most recent call last):
      File "/root/sos-master-1807/sos/sosreport.py", line 964, in setup
        plug.setup()
      File "/root/sos-master-1807/sos/plugins/sapnw.py", line 134, in setup
        self.collect_list_instances()
      File "/root/sos-master-1807/sos/plugins/sapnw.py", line 40, in collect_list_instances
        p = open(inst_out, "r").read().splitlines()
    TypeError: coercing to Unicode: need string or buffer, dict found

File: sos/plugins/abrt.py (hunk 1 above)
Comment (TurboTurtle) [reply sos/plugins/abrt.py]: Yep, oversight on my part - should be `abrt_list['status'] == 0` here.

File: sos/plugins/block.py (hunk 2 above)
Comment (TurboTurtle) [reply sos/plugins/block.py]: I'm definitely in the "always check status == 0" camp.

We have no other way to gate against the iterations that follow from executing against error output when the command fails for one reason or another.

File: sos/plugins/dnf.py (hunk 5 above)
Comment (TurboTurtle) [reply sos/plugins/dnf.py]: Fair point, will update

File: sos/plugins/openstack_glance.py (hunk 13 above)
Comment (TurboTurtle) [reply sos/plugins/openstack_glance.py]: Typo, yep - thanks for catching that.

File: sos/plugins/pcp.py (hunk 14 above)
Comment (TurboTurtle) [reply sos/plugins/pcp.py]: We previously ran `add_cmd_output('pcp')`, which would be redundant with the new approach of `get_cmd_output_now()`. So  `save=False` would be a regression here.

File: sos/plugins/vdsm.py (hunk 18 above)
Comment (TurboTurtle) [reply sos/plugins/vdsm.py]: Collecting the connected storage pools output would be beneficial to RH support teams

File: sos/plugins/vdsm.py (hunk 19 above)
Comment (TurboTurtle) [reply sos/plugins/vdsm.py]: Same here, this is useful data for support teams

File: sos/plugins/virsh.py (hunk 21 above)
Comment (TurboTurtle) [reply sos/plugins/virsh.py]: Same as `pcp` - without `save=False` we "automatically" get the `add_cmd_output()` call from above.

File: sos/plugins/virsh.py (hunk 22 above)
Comment (TurboTurtle) [reply sos/plugins/virsh.py]: Same here

File: sos/plugins/xen.py (hunk 23 above)
Comment (TurboTurtle) [reply sos/plugins/xen.py]: Ack

File: sos/plugins/yum.py (hunk 24 above)
Comment (TurboTurtle) [reply sos/plugins/yum.py]: Ack

**Comment by TurboTurtle** [comment]
After talking with Bryn a bit further, I've updated this PR to also include a rename of the `get_cmd_output_now()` methods to `collect_cmd_output()`, which `add_cmd_output()` still leverages.

Also, dropped the `save` kwarg in favor of a replacement `exec_cmd()` method to replace the `*_ext_prog()` methods.


Looking at the LGTM analysis, I'm not sure where the build failure is coming from.

**Pull request closed without merging**